WebDriver BiDi: support for navigation-related events in `browsingContext`

AnujPanta1 opened the pull request:
#### 42dcbccfce605543be196fbebde2d71a583efe62
<pre>
WebDriver BiDi: support for navigation-related events in `browsingContext`
<a href="https://bugs.webkit.org/show_bug.cgi?id=297278">https://bugs.webkit.org/show_bug.cgi?id=297278</a>
<a href="https://rdar.apple.com/137019783">rdar://137019783</a>

Reviewed by BJ Burg.

Navigation events are now hooked into WebKit&apos;s existing page load lifecycle callbacks:
didCommitLoadForFrame() for committed navigation,
didFailProvisionalLoadForFrameShared() and didFailLoadForFrame() for failed events,
didCancelClientRedirectForFrame() for aborted navigation,
didSameDocumentNavigationForFrame() variants for fragment navigation.

The implementation extends the initial navigationStarted foundation. A new helper
convertNavigationIDToString() added.

* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
(WebKit::BidiBrowsingContextAgent::navigate):
(WebKit::BidiBrowsingContextAgent::reload):
* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::navigationIDToProtocolString):
(WebKit::WebAutomationSession::navigationStartedForFrame):
(WebKit::WebAutomationSession::navigationCommittedForFrame):
(WebKit::WebAutomationSession::navigationFailedForFrame):
(WebKit::WebAutomationSession::navigationAbortedForFrame):
(WebKit::WebAutomationSession::fragmentNavigatedForFrame):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::activeAutomationSession const):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didCancelClientRedirectForFrame):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::didFailLoadForFrame):
(WebKit::WebPageProxy::didSameDocumentNavigationForFrame):
(WebKit::WebPageProxy::didSameDocumentNavigationForFrameViaJS):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/299095@main">https://commits.webkit.org/299095@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1567365801d80bb7553d2f722980d602cbd647f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117689 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37367 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27995 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123823 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69702 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89319 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/51726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30298 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105526 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69810 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29361 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67481 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99715 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23822 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126916 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44592 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33560 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97975 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44950 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101753 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97762 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24896 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43148 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21109 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40957 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44464 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50138 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43922 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47269 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45613 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a9cb9e78eda3580235bbedcb8584a983046f7494. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49280

burg commented:
```
/Volumes/Data/worker/iOS-18-Build-EWS/build/Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:949:2: error: unterminated conditional directive
  949 | #if ENABLE(WEBDRIVER_BIDI)
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/81d3c75494e5a43e420440f32452723233ea7106. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49280

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -176,6 +176,50 @@
+            "spec": "https://w3c.github.io/webdriver-bidi/#event-browsingContext-navigationCommitted",
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
```
It seems weird to me that a navigation can be nullable for a committed navigation. I looked into this and in the [introduction](https://w3c.github.io/webdriver-bidi/#module-browsingContext) to the `browsingContext` module it says:

> The [navigation id](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigation-id) for the navigation, or null when the navigation is canceled before making progress.

In order to make this more understandable on our end, I prefer we make `navigation` arguments non-`nullable` unless it can actually be null.

burg replied:
Nit: consider renaming the `Navigation` type to `NavigationID` so it isn't ambiguous, or copy the `description` from other `navigation` parameters above.

I looked further up in this file, and we should fix up all `browsingContext.Navigation` parameters while we are here. For example, this return parameter appears multiple times:

```
{ "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "optional": true, "description": "The navigation ID, or null if not applicable." }
```

I think we should make this consistent- can you update the `optional` to `nullable` in those cases?

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -176,6 +176,50 @@
+            "spec": "https://w3c.github.io/webdriver-bidi/#event-browsingContext-navigationFailed",
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
```
For the Failed and Aborted events please note what `null` means for `navigation` since it's meaningful in these events.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -176,6 +176,50 @@
+            "spec": "https://w3c.github.io/webdriver-bidi/#event-browsingContext-navigationAborted",
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext" },
+                { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "nullable": true },
```
Ditto

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7517,8 +7542,15 @@ void WebPageProxy::didFailLoadForFrame(IPC::Connection& connection, FrameIdentif
+    if (m_controlledByAutomation) {
+        if (RefPtr automationSession = m_configuration->processPool().automationSession()) {
+            auto timestamp = WallTime::now().secondsSinceEpoch().milliseconds();
+            automationSession->navigationFailedForFrame(*frame, frame->url().string(), navigationID, timestamp);
```
Since we are about to add a lot of hooks in this area of the codebase, I'm going to get more nitpicky than usual so that we can reduce future back-and-forth style comments and have nice code.

1. There is some tension between providing the `frame` object itself (because the WebAutomationSession method can check the frame URL itself and this code can become very simple) and pulling out each data member and passing them as method arguments (easier to understand what data is actually used, but now you need to maintain more automation code in WebPageProxy). In this case, I think the URL should be a `URL` type or omitted. The timestamp could be pushed into the WebAutomationSession since it doesn't use any data from this class. So then we just have `navigationFailedForFrame(*frame, navigationID)`.

1. We should try to keep this to one level of `if`, to keep our hooks as simple as possible. Option 1:

```
if (RefPtr automationSession = activeAutomationSession())
// do stuff
```

...and inside the `activeAutomationSession()` private method we do the check on `m_controlledByAutomation` and  do the pointer-chasing to get the session object.

Option 2: 
Introduce a lambda-like helper, `performIfControlledByAutomation()`:

```
performIfControlledByAutomation([](Ref session) {
// do stuff
})

```
An advantage to option 2 is that the `session` is guaranteed non-null and we can connect different lambda helpers together. But for now, option 1 should be fine.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0fd56ab06c7914dab649ed903ef418f3263158c8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49280

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2522b899e3e7a0eee175d9a55937f8854e825428. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49280

burg commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7626,6 +7649,12 @@ void WebPageProxy::didSameDocumentNavigationForFrameViaJS(IPC::Connection& conne
     protectedPageLoadState->commitChanges();
+#if ENABLE(WEBDRIVER_BIDI)
+    if (RefPtr automationSession = activeAutomationSession()) {
+        std::optional<WebCore::NavigationIdentifier> navID = navigation ? std::optional(navigation->navigationID()) : std::nullopt;
```
This local variable seems unnecessary, since the type of the parameter _should_ be enough for the compiler to know to insert implicit constructors like std::optional(T). you should be able to inline the ternary expression, omit the std::optional constructor, use initializer braces, and/or use `auto` type to make it shorter.

burg commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2795,6 +2796,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 private:
     WebPageProxy(PageClient&, WebProcessProxy&, Ref<API::PageConfiguration>&&);
     void platformInitialize();
+    RefPtr<WebAutomationSession> activeAutomationSession() const;
```
Please move this declaration down to other inspector/webdriver related method declarations.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1567365801d80bb7553d2f722980d602cbd647f4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49280

AnujPanta1 committed:
WebDriver BiDi: support for navigation-related events in `browsingContext`
https://bugs.webkit.org/show_bug.cgi?id=297278
rdar://137019783

Reviewed by BJ Burg.

Navigation events are now hooked into WebKit's existing page load lifecycle callbacks:
didCommitLoadForFrame() for committed navigation,
didFailProvisionalLoadForFrameShared() and didFailLoadForFrame() for failed events,
didCancelClientRedirectForFrame() for aborted navigation,
didSameDocumentNavigationForFrame() variants for fragment navigation.

The implementation extends the initial navigationStarted foundation. A new helper
convertNavigationIDToString() added.

* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
(WebKit::BidiBrowsingContextAgent::navigate):
(WebKit::BidiBrowsingContextAgent::reload):
* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::navigationIDToProtocolString):
(WebKit::WebAutomationSession::navigationStartedForFrame):
(WebKit::WebAutomationSession::navigationCommittedForFrame):
(WebKit::WebAutomationSession::navigationFailedForFrame):
(WebKit::WebAutomationSession::navigationAbortedForFrame):
(WebKit::WebAutomationSession::fragmentNavigatedForFrame):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::activeAutomationSession const):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didCancelClientRedirectForFrame):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::didFailLoadForFrame):
(WebKit::WebPageProxy::didSameDocumentNavigationForFrame):
(WebKit::WebPageProxy::didSameDocumentNavigationForFrameViaJS):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/299095@main
